discoverx/explorer.py

```diff
@@ -0,0 +1,287 @@
+            sql_query_template += f"TABLESAMPLE ({sample_size} ROWS)"
+
+        return self.with_sql(sql_query_template)
+
+
+class DataExplorerActions:
+    def __init__(
+        self, data_explorer: DataExplorer, spark: SparkSession = None, info_fetcher: InfoFetcher = None
+    ) -> None:
+        self._data_explorer = data_explorer
+        if spark is None:
+            spark = SparkSession.builder.getOrCreate()
+        if info_fetcher is None:
+            info_fetcher = InfoFetcher(spark)
+        self._info_fetcher = info_fetcher
+        self._spark = spark
+
+    @staticmethod
+    def _get_stack_string_columns_expression(table_info: TableInfo) -> str:
+        string_col_names = [c.name for c in table_info.columns if c.data_type.lower() == "string"]
+        stack_parameters = ", ".join([f"'{c}', `{c}`" for c in string_col_names])
+        return f"stack({len(string_col_names)}, {stack_parameters})"
+
+    @staticmethod
+    def _build_sql(sql_template: str, table_info: TableInfo) -> str:
+        if table_info.catalog and table_info.catalog != "None":
+            full_table_name = f"{table_info.catalog}.{table_info.schema}.{table_info.table}"
+        else:
+            full_table_name = f"{table_info.schema}.{table_info.table}"
+
+        stack_string_columns = DataExplorerActions._get_stack_string_columns_expression(table_info)
+
+        sql = sql_template.format(
+            table_catalog=table_info.catalog,
+            table_schema=table_info.schema,
+            table_name=table_info.table,
+            full_table_name=full_table_name,
+            stack_string_columns=stack_string_columns,
+        )
+        return sql
+
+    def _run_sql(self, sql: str, table_info: TableInfo) -> DataFrame:
+        logger.debug(f"Running SQL query: {sql}")
+
+        try:
+            df = (
+                self._spark.sql(sql)
+                .withColumn("table_catalog", lit(table_info.catalog))
+                .withColumn("table_schema", lit(table_info.schema))
+                .withColumn("table_name", lit(table_info.table))
+            )
+            logger.debug(f"Finished running SQL query: {sql}")
+            return df
+        except Exception as e:
+            logger.error(f"Error running SQL query for: {table_info.catalog}.{table_info.schema}.{table_info.table}.")
+            logger.error(e)
+            return None
+
+    def _get_sql_commands(self, data_explorer: DataExplorer) -> list[tuple[str, TableInfo]]:
+        logger.debug("Launching lakehouse scanning task\n")
+
+        table_list = self._info_fetcher.get_tables_info(
+            data_explorer._catalogs, data_explorer._schemas, data_explorer._tables, data_explorer._having_columns
+        )
+        sql_commands = [
+            (DataExplorerActions._build_sql(data_explorer._sql_query_template, table), table) for table in table_list
+        ]
+        return sql_commands
+
+    def explain(self) -> None:
+        data_explorer = self._data_explorer
```

Comment:
I would prefer to keep the reference to self._data_explorer. It makes it clear everywhere that we refer to an attribute.

Reply:
Ok, done.